### PR TITLE
[POWEROPS-2414] feat: dms updates

### PR DESCRIPTION
**Pull request by nodegard**
## Description
Please describe the change you have made.

## Checklist:
- [ ] Tests added/updated.
- [ ] Documentation updated.
- [ ] Changelog updated in [CHANGELOG.md](https://github.com/cognitedata/power-ops-sdk/blob/master/CHANGELOG.md).
- [ ] Version bumped. If triggering a new release is desired, bump the version number in [pyproject.toml](https://github.com/cognitedata/power-ops-sdk/blob/master/pyproject.toml) per [semantic versioning](https://semver.org/).

### CODE REVIEW

File: cognite/powerops/custom_modules/power_model_v1/data_models/containers/BidDocument.container.yaml
Comment (eriklien) [inline cognite/powerops/custom_modules/power_model_v1/data_models/containers/BidDocument.container.yaml:63]: Do we still have priceArea because this container is used across BidDocumentDayAhead (does not have priceArea, because it gets that from the bidConfiguration) and BidDocumentAFRR (or what it's called - probably has priceArea, but I didn't bother to check)? It doesn't really seam meaningful to me to have these two in the same container. The (not very strong?) reason for having these implement the same interface (BidDocument) is standardisation - not that we may be interested in listing both kinds of BidDocuments together (at least I don't see any case where this is relevant/meaningful).

File: cognite/powerops/custom_modules/power_model_v1/data_models/containers/DateSpecification.container.yaml
Comment (eriklien) [inline cognite/powerops/custom_modules/power_model_v1/data_models/containers/DateSpecification.container.yaml:49]: As I said earlier, I don't think this is flexible enough. I don't think this would allow us to cover all of these examples (only one of A or B, and maybe C but not D):
````
import arrow
input_time = arrow.get("2024-05-26T09:00:00+02:00")

print(input_time.floor("week").shift(days=1)) # (A) Result: 2024-05-21T00:00:00+02:00
print(input_time.shift(days=1).floor("week")) # (B) Result: 2024-05-27T00:00:00+02:00

print(input_time.shift(days=3).floor("week"))               # (C) Result: 2024-05-27T00:00:00+02:00
print(input_time.shift(days=1).floor("week").shift(days=2)) # (D) Result: 2024-05-29T00:00:00+02:00
```

File: cognite/powerops/custom_modules/power_model_v1/data_models/containers/PriceArea.container.yaml
Comment (eriklien) [inline cognite/powerops/custom_modules/power_model_v1/data_models/containers/PriceArea.container.yaml:26]: Not related to this PR, but capacityPriceUp, capacityPriceDown, activationPriceUp, activationPriceDown, relativeActivation, totalAllocationUp and totalAllocationDown should be per market/product (mFRR, aFRR, FCR-N D-1 and D-2, FCR-D D-1 and D-2), so I think we should remove these from the PriceArea container (and maybe have them in a separate PriceAreaAFRR container for now and add more later) - if they were to live in a broad, general PriceArea container, that would have to be apacityPriceUpAFRR etc.

File: cognite/powerops/custom_modules/power_model_v1/data_models/containers/ShopModel.container.yaml
Comment (eriklien) [inline cognite/powerops/custom_modules/power_model_v1/data_models/containers/ShopModel.container.yaml:55]: I thought we needed this - is it replaced by a new type, or don't we need it? As far as I know, this is something else than the shopFiles on ShopCase (which point to specific files) - cogShopFilesConfig may contain information about how to find/pick the right file (like the water value cut files, for which there are several new files every week).

File: cognite/powerops/custom_modules/power_model_v1/data_models/containers/ShopScenario.container.yaml
Comment (eriklien) [inline cognite/powerops/custom_modules/power_model_v1/data_models/containers/ShopScenario.container.yaml:47]: Why is this commented out? Not needed in container because it is a list of edges?

**Review by eriklien — COMMENTED** [review]
Reviewed containers and datamodels (will look at views and GQL afterwards).
Generally looks good, but I have added some comments/requests that I think we should address.

File: cognite/powerops/custom_modules/power_model_v1/data_models/containers/ShopCommands.container.yaml
Comment (nodegard) [reply cognite/powerops/custom_modules/power_model_v1/data_models/containers/ShopCommands.container.yaml:18]: It's the default value when the type is `text` and a change in the Python SDK made this into a required field to be provided explicitly even though you just want the default value. So it's not actually changing anything in the DMS, it's just alignment with new SDK requirements.

File: cognite/powerops/custom_modules/power_model_v1/data_models/containers/ShopModel.container.yaml
Comment (nodegard) [reply cognite/powerops/custom_modules/power_model_v1/data_models/containers/ShopModel.container.yaml:55]: It's been switched to a `multi_edge_connection` of ShopFile so the property definition has been moved to the view. I synced with Julia about it, and the contents of the JSON needed to be the same as what is defined in the ShopFile object, so I just made `cogShopFilesConfig a list of ShopFile` instead.

Comment (nodegard) [reply cognite/powerops/custom_modules/power_model_v1/data_models/containers/ShopModel.container.yaml:55]: Also, this change is detailed in the CHANGELOG, so it would be good if you referred to that as well to understand the implication of the changes.

File: cognite/powerops/custom_modules/power_model_v1/data_models/containers/DateSpecification.container.yaml
Comment (nodegard) [reply cognite/powerops/custom_modules/power_model_v1/data_models/containers/DateSpecification.container.yaml:49]: I don't think we want to support all of those things because in my opinion half of those don't provide a consistent behavior.

(A) will **always** give you the Tuesday of the "current" week given a Mon-Sun week
(B) will **either** give you the "current" week or next weeks Monday depending on what the current day is (if it's a Sunday you get next Monday if it's Monday-Saturday you get this Monday)
(C) will **either** give you the current Monday or next week's Monday depending on what day of the week you land on in 3 days
(D) will **either** give you the current Wednesday or next week's Wednesday depending on what day of the week is in 3 days

We shouldn't allow for that kind of ambiguity which is too easy to mess up right now with the current implementation. I can't see a use case in which we want an **either** situation depending on what day of the week it is?

Comment (nodegard) [reply cognite/powerops/custom_modules/power_model_v1/data_models/containers/DateSpecification.container.yaml:49]: We should only allow for non-ambiguous statements like these for example:

- this week's Monday
- yesterday
- 2 days from now
- 2 weeks from now
- Wednesday 2 weeks from now
- a month ago from today
- Thursday last week
....

It's easy to ensure this non-ambiguity by _always_ doing floor and then shift

Comment (eriklien) [reply cognite/powerops/custom_modules/power_model_v1/data_models/containers/DateSpecification.container.yaml:49]: No, I disagree. It should be possible to specify something like "The start of the week that tomorrow is in" (shift 1 day, then floor week)

Comment (eriklien) [reply cognite/powerops/custom_modules/power_model_v1/data_models/containers/DateSpecification.container.yaml:49]: And it's not ambiguity, it's just a specification for different things depending on which day of the week we are in (so is `floor("week")` - it will shift by 4.5 days from Friday noon, or by 0.25 days from Monday 06:00...)

File: cognite/powerops/custom_modules/power_model_v1/data_models/containers/ShopModel.container.yaml
Comment (eriklien) [reply cognite/powerops/custom_modules/power_model_v1/data_models/containers/ShopModel.container.yaml:55]: So we will use a ShopFile without fileReference for this?

File: cognite/powerops/custom_modules/power_model_v1/data_models/containers/DateSpecification.container.yaml
Comment (eriklien) [reply cognite/powerops/custom_modules/power_model_v1/data_models/containers/DateSpecification.container.yaml:49]: We are not building a product for only the configurations we have now.

Comment (eriklien) [reply cognite/powerops/custom_modules/power_model_v1/data_models/containers/DateSpecification.container.yaml:49]: I see the trade-off between easy to understand and flexible, but I want the flexibility here. If make a type that support this flexibility, we can make it easy to understand by including some instances with names (or external IDs?) like `tomorrow`, `end_of_next_week` etc. (which we can create as a standard part of resync or something).

File: cognite/powerops/custom_modules/power_model_v1/data_models/containers/ShopModel.container.yaml
Comment (Juliamg) [reply cognite/powerops/custom_modules/power_model_v1/data_models/containers/ShopModel.container.yaml:55]: Sorry this is my fault. I told Nina to change it to use a list of ShopFiles as input to the preprocessor as well, but since we do not know the exact `water_value_cut_file` to use before doing a lookup in CDF based on the latest ingested file, we should keep the list of json for the cogShopFilesConfig. I was looking at the file we had in powerops-staging that references the cut file directly, but this is not the case in customer environments. Example:


``` 
 - label: "water_value_cut_file"
    pick: "latest_before"  # pick the file that is closest before the start time to the triggering event
    sort_by:
      metadata_key: "update_datetime"
      file_attribute:
    external_id_prefix: "SHOP_{WATERCOURSE}_water_value_cut_file"
    file_type: "ascii"
```

File: cognite/powerops/custom_modules/power_model_v1/data_models/containers/DateSpecification.container.yaml
Comment (nodegard) [reply cognite/powerops/custom_modules/power_model_v1/data_models/containers/DateSpecification.container.yaml:49]: Resolved from the conversation that `shift(weekday=0)` accounts for what is needed in terms of flexibility

File: cognite/powerops/custom_modules/power_model_v1/data_models/containers/PriceArea.container.yaml
Comment (eriklien) [reply cognite/powerops/custom_modules/power_model_v1/data_models/containers/PriceArea.container.yaml:26]: It would also be great if priceScenarios and mainPriceScenario and defaultBidConfiguration got dayAhead as part of their name - in the container, but but in the PriceAreaDayAhead view we could skip the dayAhead part in property names)

**Review by eriklien — COMMENTED** [review]
Looked through all views in `all_PowerOps` on space `nina_sp_power_ops_models` on `power-ops-staging`. Here are my comments

I guess comments/changes to descriptions can wait?

Alert (I think we've already discussed that we may want to do some changes here later, and can keep it as is for now, but here are some remarks anyways)
- EventIds has to be deprecated
- What is calculationRun / how should it be used / what should it be used for?
- alertType+statusCode+severity seems a bit too much - should we get rid of one of them or redesign this?

ShopTriggerOutput
- Does it make sense to include shopResult here (which is not the output of the trigger, but of the actual run of CogShop)? Will CogShop add this to the ShopTriggerOutput?


TotalBidMatrixCalculationInput
- We've realised that this one doesn't contain everything we need to create a BidDocument... Two options forward (that we probably should wait with):
1. Change TotalBidMatrixCalculationInput such that we just return the total bid matrix sequence and post-processed partials (with some reference to the initial ones, such that we know which is for what), and create BidDocument outside of this "logical function" (but still have it in the Cognite function)
2. Include PartialBidMatrixCalculationOutputs (instead of partials?) in the TotalBidMatrixCalculationInput

ShopOutputTimeSeriesDefinition
- [as commented on the container] Unit could be unit_external_id (for unit conversion), but maybe leave that for later.

FunctionOutput (and all derived types)
- input: Should we consider renaming this to e. g. functionInput to avoid "conflict" with keyword "input" (pygen handles this by making this "input_")

PartialBidMatrixCalculationOutput
- partialMatrix - rename to partialBidMatrix? (Can check what we have done elsewhere)

Generator:
- [same comment as on container]: Rename startStopCost to startCostTimeSeries (or to startStopCostTimeSeries and rename startCost to startStopCost)
- why "@name turbineCurves" before turbineEfficiencyCurves? (just for my understanding)

Watercourse, Plant and PriceArea:
- We should consider modelling the link between plants and price areas and watercourses (but that may also wait)

PriceAreaDayAhead
- [as mentioned elsewhere]: May be good to add actualDayAheadPrice (or actualPrice if we do not add "DayAhead" to defaultBidConfiguration, mainPriceScenario and priceScenarios)

PriceAreaInformation
- See comment on container


BidDocumentDayAheadSimple
- I don't remember what the purpose of this one is - do you?

ShopPenaltyReport
- Would probably be good to link this one to a ShopResult. Not urgent (and we already have ShopResult.alerts)

File: cognite/powerops/custom_modules/power_model_v1/data_models/containers/ShopModel.container.yaml
Comment (nodegard) [reply cognite/powerops/custom_modules/power_model_v1/data_models/containers/ShopModel.container.yaml:55]: Resolved with @Juliamg by adding an optional `external_id_prefix` field to ShopFile

File: cognite/powerops/custom_modules/power_model_v1/data_models/containers/ShopAttributeMapping.container.yaml
Comment (nodegard) [reply cognite/powerops/custom_modules/power_model_v1/data_models/containers/ShopAttributeMapping.container.yaml]: Updated in functions to be case-insensitive so using uppercase for all defaults

**Comment by nodegard** [comment]
- Alert: hold off on changes
- ShopTriggerOutput: needs to be aligned with rest of team, recommendation is to hold off
- TotalBidMatrixCalculationInput: hold off on changes
- ShopOutputTimeSeriesDefinition: added TODO and hold off on it until further alignment and testing
- FunctionOutput: done changed input to functionInput
- PartialBidMatrixCalculationOutput: suggest holding off on this as it's only a view change not container, and other places we use just "partials"
- Generator:
- 
- Watercourse, Plant and PriceArea

File: cognite/powerops/custom_modules/power_model_v1/data_models/containers/BidDocument.container.yaml
Comment (nodegard) [reply cognite/powerops/custom_modules/power_model_v1/data_models/containers/BidDocument.container.yaml:63]: I'd say both types of bid documents contain all the same fields, but the only difference is what type a few of the direct relations/multi_edge_connections point to hence why I think it's fair to have it all in one container.

Day Ahead has relations to `bidConfiguration`, `total`, and `partials`. While current AFFR (which probably will need some changes when we actually start on AFFR) has relations to `priceArea` (would think this maybe gets replaced with a similar "configuration" object like we have for Day Ahead) and `bids` (BidRow which to me is "the same" as the relation to BidMatrix for DayAhead)

**Review by eriklien — COMMENTED** [review]
Nice and fast progress! Let me know when you want a new review/approval.

File: cognite/powerops/custom_modules/power_model_v1/data_models/containers/PriceArea.container.yaml
Comment (eriklien) [reply cognite/powerops/custom_modules/power_model_v1/data_models/containers/PriceArea.container.yaml:26]: I think this one is resolved through what we discussed today (one instance/row in the container per market, and assume actual price is part of mainPriceScenario) 👍

**Review by eriklien — COMMENTED** [review]
New changes look good. This is the only remaining thing I am aware of:
```
AssertionError: Version in pyproject.toml (0.94.5) does not match version in cognite/powerops/version.py (0.94.5) or in the CHANGELOG.md 0.94.6
assert '0.94.5' == '0.94.6'
```

**Comment by nodegard** [comment]
All changes should be done now, resync populated data can be validated [here](https://cog-power-ops.fusion.cognite.com/power-ops-staging/data-models/power_ops_core/all_DayAhead/1/data-management/preview?type=BidConfigurationDayAhead&cluster=bluefield.cognitedata.com&workspace=data-management&search=&sort=&order=&filter=&advancedFilter=)

**Review by eriklien — APPROVED** [review]
Looking good, and thanks for running resync again.
There are some updates we should do for resync data, but I/we can do that in a separate PR.

**Review by Juliamg — APPROVED** [review]

**Pull request merged**
